scene/main/scene_tree.cpp

```diff
@@ -122,27 +122,45 @@ void SceneTree::tree_changed() {
 }
 
 void SceneTree::node_added(Node *p_node) {
-	emit_signal(node_added_name, p_node);
-	if (call_lock > 0) {
-		call_skip.erase(p_node->get_instance_id());
+	ERR_FAIL_NULL(p_node);
+	// Update `call_skip` before emitting the signal so it will be up to date for anything trigerred by the signal.
+	if (call_skip) {
+		// Remove the pointer from `call_skip` only if it still points to the same node that was removed. It allows
+		// such node to still be iterated over during the ongoing group call.
+		// In case it's a new node that was allocated at the address of a previously freed node which were meant to
+		// be skipped, we still want to skip such pointer because during the ongoing group call this pointer refers
+		// to that already freed node.
+		ObjectID *removed_node_id = call_skip->getptr(p_node);
+		if (removed_node_id && (*removed_node_id) == p_node->get_instance_id()) {
+			call_skip->erase(p_node);
+		}
 	}
+	emit_signal(node_added_name, p_node);
 }
 
 void SceneTree::node_removed(Node *p_node) {
+	ERR_FAIL_NULL(p_node);
 	if (current_scene == p_node) {
 		current_scene = nullptr;
 	}
-	emit_signal(node_removed_name, p_node);
-	if (call_lock > 0) {
-		call_skip.insert(p_node->get_instance_id());
+	// Update `call_skip` before emitting the signal so it will be up to date for anything trigerred by the signal.
+	if (call_skip) {
+		// Store the pointer to be skipped (and the node's ID) only if it's not yet meant to be skipped. Otherwise
+		// it's pointing to a new node allocated at the pointed address and thus we don't want to overwrite the ID
+		// of the node which was originally meant to be skipped (and which must have been freed by now).
+		if (!call_skip->has(p_node)) {
+			call_skip->insert(p_node, p_node->get_instance_id());
+		}
```

Comment:
Can this "insert if not yet existing" could be done without a duplicated lookup (`HashMap::insert` does the lookup for the existing entry anyway)? :thinking: That's no really an issue though, just wondering.